src/array/range.ts

```diff
@@ -0,0 +1,48 @@
+/**
+ * Returns an array of numbers from `start` to `end` with the specified `step` value.
+ *
+ * This function generates a range of numbers, starting from `start`, ending before `end`,
+ * and incrementing by `step`. If `step` is not provided, it defaults to `1` for an
+ * ascending range and `-1` for a descending range.
+ *
+ * @param {number} start - The starting number of the range.
+ * @param {number} [end] - The end number of the range.
+ * @param {number} [step] - The step value for the range.
+ * @returns {number[]} An array of numbers from `start` to `end` with the specified `step`.
+ *
+ * @example
+ * // Returns [0, 1, 2, 3]
+ * range(4);
+ *
+ * @example
+ * // Returns [0, 5, 10, 15]
+ * range(0, 20, 5);
+ *
+ * @example
+ * // Returns [0, -1, -2, -3]
+ * range(0, -4, -1);
+ *
+ * @example
+ * // Returns [1, 1, 1]
+ * range(1, 4, 0);
+ */
+
+export function range(start: number, end?: number, step?: number): number[] {
+  if (end === undefined) {
+    end = start;
+    start = 0;
+  }
+
+  if (step === undefined) {
+    step = start <= end ? 1 : -1;
+  }
```

Comment:
I think we should return an empty array if `end` is smaller than `start`.

For example, python returns an empty array when `end` is smaller than `start`.

<img width="191" alt="image" src="https://github.com/toss/es-toolkit/assets/3102175/162c8858-25ab-4730-83db-f38dc8c77867">

Reply:
lodash also supports the end is smaller than start case, as I mention previous comment, it would help lodash user to migrate to es-toolkit.

Reply:
I see why we have this behavior here. However, to keep our implementation simple, I prefer not to support this use case. 

It might make it harder for people using lodash to migrate, but I don't think this is a major use case.